002-tidyr-stringr.Rmd

```diff
@@ -0,0 +1,256 @@
+Unfortunately, these functions are only weakly connected to their stringr equivalents, making them harder to learn and remember.
+This tidyup proposes a new cohesive family of string manipulation functions to tackle common problems that arise during data tidying.
+The existing functions will remain so existing code is unaffected, but will be superseded to steer folks towards a more consistent API.
+
+```{r setup}
+library(tidyr)
+library(stringr)
+```
+
+## Motivation
+
+stringr and tidyr provide surprisingly different interfaces for string manipulation.
+This means that common tasks require integrating two distinct syntaxes and learning more about stringr doesn't help your use of tidyr.
+
+The root cause of the problem is that stringr was designed before I understood the importance of data frames, so stringr functions that need to return multiple values use matrices rather than data frames.
+This is particularly unfortunate given that `dplyr::summarse()` and `mutate()` can now create multiple columns from data frame results.
+
+To overcome this weakness, tidyr provides family of string manipulation functions that work with data frames.
+Unfortunately, the names of the tidyr functions have little relationship to their stringr equivalents:
+
+-   `extract()` creates new variables from groups, and is similar to `str_match()`.
+-   `separate(sep = "pattern")` creates new variables by splitting with a regular expression, and is similar to `str_split_fixed()`.
+-   `separate(sep = c(2, 5, 10)`) creates new variables by splitting based on position, and has no equivalent in stringr.
+-   `separate_rows()` creates new rows, and is similar to `str_split()`.
+
+Additionally, tidyr uses the base R regexp engine and syntax, not the stringr engine and syntax.
+There are relatively few differences between the engines, but the syntax for controlling regular expression options (e.g. ignoring case) is very different.
+In the past, we have kept stringr out of tidyr dependencies because it can be expensive/painful to build stringi on linux.
+However, most tidyverse users will have stringr installed, and RSPM makes stringi easier to install, so it would be nice to use it if available, following the footsteps laid by [Radix Ordering in](https://github.com/tidyverse/tidyups/blob/main/003-dplyr-radix-ordering.md) [`dplyr::arrange()`](https://github.com/tidyverse/tidyups/blob/main/003-dplyr-radix-ordering.md).
+
+Additionally there are a few gaps in the current API:
+
+-   Neither stringr nor tidyr provide a particularly nice way to work with named capture groups.
+    There's also scope for more directly matching column names to patterns if `extract()` didn't attempt to match the syntax of `separate()` so closely, e.g. `str_extract(x, c(type = ".*", "|", fruit = "orange|apple"))`.
+
+-   tidyr only provides one function for splitting a string into rows, not columns: `separate_rows()`.
+    What other tidyr functions should create rows?
```

Comment:
Ok I see you have the `Eliminate tidyr functions` section below, but maybe we can think about it one more time with `unchop()`

Reply:
I think the well-named tidyr helpers are useful, even if the same task can be accomplished with dplyr+stringr, particularly for common tasks, and I think the examples here are common data wrangling tasks. Some of the dplyr+stringr combinations can be difficult to learn early on because you have to think both across columns and down the rows at the same time to understand what the code is doing and what results to expect.

Reply:
I think I am mainly excited about the fact that we can accomplish this with dplyr + `unchop()` / `unpack()` at all.

AFAIK, we previously didn't think we had an existing solution for what `str_*_longer()` + dplyr + `unchop()` would do - so now that we do have a way to perform that operation it is at least clear what the tidyr implementation of things like `separate_by_longer()` will be. I agree that this combination could be difficult for new R users to come up with on their own, so it is probably fine for the tidyr variants to exist.

I was a little excited about the idea that tidyr could get out of the string manipulation business entirely, since `separate()` and `extract()` never felt quite right in tidyr, but that is probably overambitious 😬

Reply:
Yeah, I think the main take-away is that the tidyr functions will be mostly thin wrappers around the stringr functions which do the hard work of implementation.